lib/api/model/model.dart

```diff
@@ -698,3 +710,74 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  // Adapted from the shared code:
+  //   https://github.com/zulip/zulip/blob/1fac99733/web/shared/src/resolved_topic.ts
+  // Pattern for an arbitrary resolved-topic prefix.
+  // These always begin with the canonical prefix, but can go on longer.
+  // It's designed to remove a weird "✔ ✔✔ " prefix, if present.
+  static RegExp _resolvedTopicPrefixRe = RegExp('^✔ [ ✔]*');
+
+  /// Whether two topics are equal, ignoring any resolved-topic prefix.
+  ///
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. Topics whose only difference is the ✔ prefix are considered
+  /// the same. This helper can be helpful when checking if a message has been
+  /// moved.
+  static bool areSameTopic(String topic, String prevTopic) {
+    // TODO(#744) Extract this to its own home to support "mark as resolve".
+    topic = topic.replaceFirst(_resolvedTopicPrefixRe, '');
+    prevTopic = prevTopic.replaceFirst(_resolvedTopicPrefixRe, '');
+
+    return topic == prevTopic;
```

Comment:
This seems to be different from the logic that's used in `analyze_edit_history` at that very helpful link you added:
https://github.com/zulip/zulip/blob/c31cebbf68a93927d41e9947427c2dd4d46503e3/web/src/message_list_view.js#L68-L118

Let's adjust this code to have the same behavior as what's in web — the question of "was this message moved or not" is basically part of the Zulip model, so it's good for different clients to agree on that question even in edge cases.

Reply:
Interestingly, the test

```
      test('Unresolving topic with a weird prefix -> none', () {
          checkEditState(MessageEditState.none,
            [{'prev_topic': '✔ ✔old_topic', 'topic': 'old_topic'}]);
      });
```

fails with the implementation change. This definitely aligns a previously different behavior [with the web app](https://chat.zulip.org/#narrow/stream/7-test-here/topic/test.20move.20weird.20prefix). I will change the expected value to `MessageEditState.moved`.